ampe.c

```diff
@@ -1065,11 +1097,29 @@ int process_ampe_frame(struct ieee80211_mgmt_frame *mgmt, int len,
         return 0;
     }
 
-    /* "1" here means only get peers in SAE_ACCEPTED */
-    if ((cand = find_peer(mgmt->sa, 1)) == NULL) {
-		sae_debug(AMPE_DEBUG_FSM, "Mesh plink: plink open from unauthed peer "MACSTR"\n",
-                  MAC2STR(mgmt->sa));
-        return 0;
+    /* require authed peers if secure mesh */
+    if (ampe_conf.mesh->conf->is_secure) {
+        /* "1" here means only get peers in SAE_ACCEPTED */
+        if ((cand = find_peer(mgmt->sa, 1)) == NULL) {
+            sae_debug(AMPE_DEBUG_FSM, "Mesh plink: plink open from unauthed peer "MACSTR"\n",
+                      MAC2STR(mgmt->sa));
+            return 0;
+        }
+    } else {
+        /*
+         * In open mesh, there's no auth stage, so we create the station
+         * when the first mgmt frame or beacon is received.  Do that now
+         * if we haven't already.
+         */
+        cand = find_peer(mgmt->sa, 0);
+        if (!cand) {
+            cand = create_candidate(mgmt->sa, me, 0, cookie);
```

Comment:
Where do we check that this beacon comes from a candidate peer mesh STA? 
Quoting from the standard... 

> 14.2.7 Candidate peer mesh STA
> When a mesh STA discovers a neighbor mesh STA through the scanning process and the discovered mesh STA is considered a candidate peer mesh STA, it may become a member of the mesh BSS of which the discovered mesh STA is a member and establish a mesh peering with the neighbor mesh STA.
> The discovered neighbor mesh STA shall be considered a candidate peer mesh STA if and only if all of the following conditions are met:
> a) The mesh STA uses the same mesh profile as the received Beacon or Probe Response frame indicates for the neighbor mesh STA.
> NOTE—If the scanning mesh STA has not become a member of any MBSS yet, it might simply activate the same mesh profile as the discovered neighbor mesh STA’s profile to fulfill this condition.
> b) The Accepting Additional Mesh Peerings subfield in the Mesh Capability field in the received Beacon or Probe Response frame equals 1.
> c) The mesh STA supports the data rates indicated by the BSSBasicRateSet of the received Beacon or Probe Response frame.
> d) If both the scanning mesh STA and the discovered neighbor STA are HT STAs, the STA has the same value in the Basic HT-MCS Set field of the HT Operation parameter of the MLME- START.request primitive as the received Beacon or Probe Response frame indicates for the neigh- bor mesh STA.
> e) If both the scanning mesh STA and the discovered neighbor STA are VHT STAs, the mesh STA uses the same value for the Basic VHT-MCS And NSS Set field in its VHT Operation element as received in the Beacon or Probe Response frame from the neighbor mesh STA.
> f) If the scanning mesh STA has dot11MeshSecurityActivated equal to true and the dot11MeshActiveAuthenticationProtocol is ieee8021x (2), either the scanning mesh STA has an active connection to an AS or the discovered mesh STA has the Connected to AS subfield in the Mesh Formation field in the Mesh Configuration element equal to 1 in the received Beacon or Probe Response frame.

Reply:
I think the kernel does this already -- mesh_neighbour_update which is where the candidate events are created -- is wrapped in a mesh_matches_local() call.  But I'll add a test case for this also: mismatched candidates should not be able to peer.